sklearn/utils/tests/test_validation.py

```diff
@@ -218,6 +219,25 @@ def test_check_array():
     assert_true(isinstance(result, np.ndarray))
 
 
+def test_check_array_pandas_dtype_object_conversion():
+    # test that data-frame like objects with dtype object
+    # get converted
+    X = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]], dtype=np.object)
+    X_df = MockDataFrame(X)
+    assert_equal(check_array(X_df).dtype.kind, "f")
+    assert_equal(check_array(X_df, ensure_2d=False).dtype.kind, "f")
+    # smoke-test against dataframes with column named "dtype"
+    X_df.dtype = "Hans"
+    assert_equal(check_array(X_df, ensure_2d=False).dtype.kind, "f")
```

Comment:
``check_array`` completely ignores the attribute. The MockDataFrame doesn't try to inspect ``self.__dict__`` so the array that is returned has nothing to do with "Hans". The issue I am smoke-testing is that ``check_array`` assumes a ``dtype`` object to be a numpy dtype and if it is not, it crashed.
The semantics I intended in this fix is that "if dtype is no vaild dtype, then this thing is probably not a numpy array and we just try to convert it".